pkg/liquidity-mining/test/GaugeWorkingBalanceHelper.test.ts

```diff
@@ -0,0 +1,300 @@
+          gauge.address,
+          user.address
+        );
+
+        // Ensure we have equal balances (that are non-zero)
+        expect(fromFp(projectedWorkingBalance) / fromFp(currentWorkingBalance)).to.eq(MAX_BALANCE_RATIO);
+        expect(projectedWorkingBalance).to.be.gt(0);
+      });
+    }
+
+    function projectedIsIncreased(maxPctIncrease: number) {
+      it(`projected balance should be slightly more than current, by less than ${maxPctIncrease}%`, async () => {
+        const [currentWorkingBalance, projectedWorkingBalance] = await workingBalanceHelper.getWorkingBalances(
+          gauge.address,
+          user.address
+        );
+
+        expect(projectedWorkingBalance).to.be.gt(currentWorkingBalance);
+
+        const ratio = fromFp(projectedWorkingBalance) / fromFp(currentWorkingBalance);
+
+        expect(ratio).to.be.lt(1 + maxPctIncrease / 100);
+      });
+    }
+
+    function currentShouldAlmostEqualProjected() {
+      it('projected balance should be close to or less than current', async () => {
+        const [currentWorkingBalance, projectedWorkingBalance] = await workingBalanceHelper.getWorkingBalances(
+          gauge.address,
+          user.address
+        );
+
+        // Ensure we have equal balances (that are non-zero)
+        expect(projectedWorkingBalance).to.be.almostEqual(currentWorkingBalance);
+        expect(projectedWorkingBalance).to.be.gt(0);
+        expect(projectedWorkingBalance).to.be.lte(currentWorkingBalance);
+      });
+    }
+
+    function itUpdatesAfterCheckpointing() {
+      context('when checkpointed', () => {
+        sharedBeforeEach('checkpoint user', async () => {
+          await gauge.connect(user).user_checkpoint(user.address);
+        });
+
+        currentShouldAlmostEqualProjected();
+      });
+    }
+
+    describe('with no veBAL', () => {
+      context('check raw balances', () => {
+        sharedBeforeEach('deposit', async () => {
+          depositIntoGauge(user, stakeAmount);
+        });
+
+        currentBalanceEqualsProjected();
+      });
+
+      context('check balance ratios', () => {
+        sharedBeforeEach('deposit', async () => {
+          depositIntoGauge(user, stakeAmount);
+        });
+
+        currentRatioEqualsProjected();
+      });
+    });
+
+    describe('with a veBAL monopoly', () => {
+      sharedBeforeEach('deposit', async () => {
+        await depositIntoGauge(user, stakeAmount);
+        await createLockForUser(user, bptAmount, LOCK_PERIOD);
+      });
+
+      projectedIsMaxedOut();
+
+      itUpdatesAfterCheckpointing();
+    });
+
+    describe('with 1% veBAL', () => {
+      sharedBeforeEach('deposit', async () => {
+        // Another person stakes 100x as much
+        await depositIntoGauge(user, stakeAmount);
+        await depositIntoGauge(other, stakeAmount);
+
+        await createLockForUser(user, bptAmount, 365 * DAY);
+        await createLockForUser(other, bptAmount.mul(100), 365 * DAY);
+      });
+
```

Comment:
These are very good, I like them a lot. I'd add one small thing: in both of these cases, have a `context` in which we do a checkpoint, advance time by some amount (say half of the lock period), and then check that the projected values (both balance and ratio) are stricly lower than the current ones (because of the veBAL decay).